k9mail-library/src/main/java/com/fsck/k9/mail/internet/MimeUtility.java

```diff
@@ -990,24 +993,37 @@ public static boolean isDefaultMimeType(String mimeType) {
         return isSameMimeType(mimeType, DEFAULT_ATTACHMENT_MIME_TYPE);
     }
 
-    public static Body createBody(InputStream in, String contentTransferEncoding, String contentType)
+    public static Body createBody(InputStream in, String contentTransferEncoding, String contentType, final AttachmentProgressCallback progressCallback)
             throws IOException {
 
         if (contentTransferEncoding != null) {
             contentTransferEncoding = MimeUtility.getHeaderParameter(contentTransferEncoding, null);
         }
 
-        BinaryTempFileBody tempBody;
+        final BinaryTempFileBody tempBody;
         if (MimeUtil.isMessage(contentType)) {
             tempBody = new BinaryTempFileMessageBody(contentTransferEncoding);
         } else {
             tempBody = new BinaryTempFileBody(contentTransferEncoding);
         }
 
         OutputStream out = tempBody.getOutputStream();
+        Timer timer = null;
         try {
+            if(progressCallback != null){
+                timer = new Timer();
+                timer.scheduleAtFixedRate(new TimerTask() {
+                    @Override
+                    public void run() {
+                        progressCallback.onUpdate((int) tempBody.getSize());
```

Comment:
There's no guarantee the file size will reflect the actual number of bytes written before the `FileOutputStream` has been closed. I'd prefer to wrap the stream in a `CountingOutputStream` and get the count from there.